Closes #7604: Add filter modifier dropdowns for advanced lookup operators

jnovinger opened the pull request:
### Closes: #7604

Implements dynamic filter modifier UI that allows users to select lookup operators (exact, contains, starts with, regex, negation, empty/not empty) directly in filter forms without manual URL parameter editing.

Supports filters for all scalar types and strings, as well as some related object filters. Explicitly does not support filters on fields that use `APIWidget`. That has been broken out in to follow up work.

#### How It Works

```
User selects "contains" ─────────────────┐
                                         │
                                         ▼
┌──────────────────────────────────────────────────────────┐
│  Filter Form (Browser)                                   │
│  ┌────────────────┐  ┌──────────────────────────┐        │
│  │ [Contains ▼]   │  │ Input: ABC               │        │
│  └────────────────┘  └──────────────────────────┘        │
│         │                                                │
│         │ TypeScript handler (filterModifiers.ts)        │
│         ▼                                                │
│  Updates input name: "serial" -> "serial__ic"            │
│  Updates URL param:  ?serial__ic=ABC                     │
└──────────────────────────────────────────────────────────┘
                         │
                         │ Form submission
                         ▼
┌──────────────────────────────────────────────────────────┐
│  Django Backend                                          │
│                                                          │
│  FilterModifierWidget                                    │
│    ↓ Renders modifier dropdown + original widget         │
│                                                          │
│  FilterModifierMixin                                     │
│    ↓ Enhances filterset fields with appropriate lookups  │
│                                                          │
│  FilterSet processes: serial__ic=ABC                     │
│    ↓ Django ORM: .filter(serial__icontains='ABC')        │
│                                                          │
│  Results + Filter Pills                                  │
│    ↓ "Serial: contains ABC [×]"                          │
└──────────────────────────────────────────────────────────┘
```

#### Details

##### Backend: `FilterModifierWidget`

Wraps any Django form widget with a modifier dropdown. Key method:

```python
def value_from_datadict(self, data, files, name):
    # Check all possible lookup variants (exact, ic, isw, n, etc.)
    for lookup_code, _ in self.lookups:
        param_name = f"{name}__{lookup_code}" if lookup_code != 'exact' else name
        if param_name in data:
            return data.get(param_name)
```

This allows the widget to find its value regardless of which lookup modifier is active in the URL.

##### Backend: `FilterModifierMixin`

Automatically enhances filterset form fields based on their type:

- **`CharField`** -> `exact`, `contains` (`ic`), `startswith` (`isw`), `endswith` (`iew`), `iexact` (`ie`), `negation` (`n`), `regex`, `iregex`, `empty`
- **`IntegerField`** -> `exact`, `gte`, `lte`, `gt`, `lt`, `negation` (`n`), `empty`
- **`DecimalField`** -> `exact`, `gte`, `lte`, `gt`, `lt`, `negation` (`n`), `empty`
- **`DateField`** -> `exact`, `gte`, `lte`, `gt`, `lt`, `negation` (`n`), `empty`
- **`ChoiceField`** -> `exact`, `negation` (`n`), `empty`
- **`MultipleChoiceField`** -> `exact`, `negation` (`n`), `empty`
- **`ModelChoiceField`** -> `exact`, `negation` (`n`), `empty`
- **`ColorField`** -> `exact`, `negation` (`n`), `empty`
- **`TagFilterField`** -> `exact`, `negation` (`n`), `empty`

##### Frontend: `filterModifiers.ts`

Typescript handler that:
1. Detects modifier selection change
2. Updates the associated input's `name` attribute
3. Syncs URL parameters when navigating with existing filters

##### Filter Pills Enhancement

Modified `applied_filters` template tag to:
1. Detect lookup modifier in parameter name (`serial__ic`)
2. Display human-readable label ("contains") instead of raw lookup code
3. Maintain correct removal URL (preserves other filters)

#### Scope & Compatibility

**What's changed:**
- All `FilterForm` subclasses now have modifier dropdowns
- Filter pills show lookup type
- URL parameters include lookup suffixes (`__ic`, `__n`, etc.)

**What's NOT changed:**
- Existing URL parameters still work (backward compatible)
- Django `FilterSet` behavior unchanged
- No database schema changes
- No API changes

jnovinger committed:
Fixes #7604: Add filter modifier dropdowns for advanced lookup operators

Implements dynamic filter modifier UI that allows users to select lookup operators
(exact, contains, starts with, regex, negation, empty/not empty) directly in filter
forms without manual URL parameter editing.

Supports filters for all scalar types and strings, as well as some
related object filters. Explicitly does not support filters on fields
that use APIWidget. That has been broken out in to follow up work.

**Backend:**
- FilterModifierWidget: Wraps form widgets with lookup modifier dropdown
- FilterModifierMixin: Auto-enhances filterset fields with appropriate lookups
- Extended lookup support: Adds negation (n), regex, iregex, empty_true/false lookups
- Field-type-aware: CharField gets text lookups, IntegerField gets comparison operators, etc.

**Frontend:**
- TypeScript handler syncs modifier dropdown with URL parameters
- Dynamically updates form field names (serial → serial__ic) on modifier change
- Flexible-width modifier dropdowns with semantic CSS classes

jnovinger committed:
Remove extraneous TS comments

jnovinger committed:
Fix import order

jnovinger committed:
Fix CircuitFilterForm inheritance

jeremystretch commented on netbox/utilities/forms/filterset_mappings.py:
```diff
@@ -0,0 +1,11 @@
+#   from .filtersets import XFilterSet
+#   FILTERSET_MAPPINGS[XFilterForm] = XFilterSet
+
+FILTERSET_MAPPINGS = {}
```
We should avoid introducing one-off globals to be used for making associations in this manner. Instead, we can leverage the [application registry](https://netboxlabs.com/docs/netbox/development/application-registry/) to correlate FilterSets to their corresponding models; the models are already set on the filter forms. Then you could do something like this:

```python
@register_filterset
class CircuitFilterSet(PrimaryModelFilterSet, TenancyFilterSet, ContactModelFilterSet):
```

```python
from netbox.registry import registry
filterset = registry['filtersets'].get(f'{app_label}.{model}')
```

We take a similar approach to map models to their search indexes (`registry['search']`).

jnovinger committed:
Enable filter form modifiers on DCIM models

jnovinger committed:
Enable filter form modifiers on Tenancy models

jnovinger committed:
Enable filter form modifiers on Wireless models

jnovinger committed:
Enable filter form modifiers on IPAM models

jnovinger committed:
Enable filter form modifiers on VPN models

jnovinger committed:
Enable filter form modifiers on Virtualization models

jnovinger committed:
Enable filter form modifiers on Circuit models

jnovinger committed:
Enable filter form modifiers on Users models

jnovinger committed:
Enable filter form modifiers on Core models

jnovinger committed:
Enable filter form modifiers on Extras models

jnovinger committed:
Add ChoiceField support to FilterModifierMixin

Enable filter modifiers for single-choice ChoiceFields in addition to the
existing MultipleChoiceField support. ChoiceFields can now display modifier
dropdowns with "Is", "Is Not", "Is Empty", and "Is Not Empty" options when
the corresponding FilterSet defines those lookups.

The mixin correctly verifies lookup availability against the FilterSet, so
modifiers only appear when multiple lookup options are actually supported.
Currently most FilterSets only define 'exact' for single-choice fields, but
this change enables future FilterSet enhancements to expose additional
lookups for ChoiceFields.

jeremystretch commented on netbox/utilities/templatetags/helpers.py:
```diff
@@ -418,22 +419,69 @@ def applied_filters(context, model, form, query_params):
+                'lt': _('<'),
+                'lte': _('≤'),
+            }
+            if modifier != 'exact' and modifier in lookup_labels:
```
Can we simplify this a bit to check for `label := lookup_labels.get(modifier)`?

jeremystretch reviewed (changes requested):
I haven't quite completed a review (still need to go through the Typescript), but wanted to deliver some suggestions and clarify some pieces. Great work!

jnovinger committed:
Address PR feedback: Replace global filterset mappings with registry

jnovinger committed:
Address PR feedback: Move FilterModifierMixin into base filter form classes

Incorporates FilterModifierMixin into NetBoxModelFilterSetForm and FilterForm,
making filter modifiers automatic for all filter forms throughout the application.

jnovinger committed:
Fix filter modifier form submission bug with 'action' field collision

Forms with a field named "action" (e.g., ObjectChangeFilterForm) were causing
the form.action property to be shadowed by the field element, resulting in
[object HTMLSelectElement] appearing in the URL path.

Use form.getAttribute('action') instead of form.action to reliably retrieve
the form's action URL without collision from form fields.

Fixes form submission on /core/changelog/ and any other forms with an 'action'
field using filter modifiers.

jnovinger committed:
Address PR feedback: Move FORM_FIELD_LOOKUPS to module-level constant

Extracts the field type to lookup mappings from FilterModifierMixin class
attribute to a module-level constant for better reusability.

jnovinger committed:
Address PR feedback: Refactor and consolidate field filtering logic

Consolidated field enhancement logic in FilterModifierMixin by:
- Creating QueryField marker type (CharField subclass) for search fields
- Updating FilterForm and NetBoxModelFilterSetForm to use QueryField for 'q'
- Moving all skip logic into _get_lookup_choices() to return empty list for
  fields that shouldn't be enhanced
- Removing separate _should_skip_field() method
- Removing unused field_name parameter from _get_lookup_choices()
- Replacing hardcoded field name check ('q') with type-based detection

jnovinger committed:
Address PR feedback: Refactor applied_filters to use FORM_FIELD_LOOKUPS

jnovinger committed:
Address PR feedback: Rename FilterModifierWidget parameter to widget

jnovinger committed:
Fix registry pattern to use model identifiers as keys

Changed filterset registration to use model identifiers ('{app_label}.{model_name}')
as registry keys instead of form classes, matching NetBox's pattern for search indexes.

jnovinger committed:
Address PR feedback: refactor brittle test for APISelect useage

Now checks if widget is actually APISelect, rather than trying to infer
from the class name.

jnovinger committed:
Refactor register_filterset to be more generic and simple

jnovinger committed:
Remove unneeded imports left from earlier registry work

jnovinger committed:
Update app registry for new `filtersets` store

jnovinger commented on netbox/utilities/forms/mixins.py:
```diff
@@ -75,3 +79,177 @@ class DistanceValidationMixin(forms.Form):
+    def _is_api_widget_field(self, field):
+        """Check if a field uses an API-based widget."""
+        # Check field class name
+        if 'Dynamic' in field.__class__.__name__:
```
Surprisingly, something like that was still required, yes. I ended up changing it to recognize the fields based by type-checking the widget. FWIW, this will be completely removed by the follow up work for API-based fields.

jeremystretch commented on netbox/utilities/forms/mixins.py:
```diff
@@ -5,13 +5,98 @@
+        ('regex', _('Matches Pattern')),
+        ('iregex', _('Matches Pattern (case-insensitive)')),
+        (MODIFIER_EMPTY_TRUE, _('Is Empty')),
+        (MODIFIER_EMPTY_FALSE, _('Is Not Empty')),
```
Very minor nit here, but IMO we shouldn't use title case for these labels

jeremystretch commented on netbox/utilities/forms/mixins.py:
```diff
@@ -75,3 +160,89 @@ class DistanceValidationMixin(forms.Form):
+
+        for field_class in field.__class__.__mro__:
+            if field_class in FORM_FIELD_LOOKUPS:
+                return FORM_FIELD_LOOKUPS[field_class]
```
One thing I've noticed is that the empty/not empty lookups are added even for mandatory model fields (e.g. `Site.status`). While not necessarily problematic, this can be confusing to the user. However, I don't see any direct path to glean that context via the model's filterset. Not sure it's worth tackling at this time.

jnovinger replied:
I took a stab at this, by trying to check the the field on form and then on the model. I didn't love it.

jnovinger committed:
Remove unused star import, leftover from earlier work

jnovinger committed:
Enables filter modifiers on APISelect based fields

jnovinger committed:
Support filter modifiers for ChoiceField

jnovinger committed:
Fix filterset registration for doubly-registered models

jnovinger commented on netbox/utilities/filtersets.py:
```diff
@@ -0,0 +1,13 @@
+from netbox.registry import registry
+
+
+def register_filterset(filterset_class):
```
Do you mean `netbox/utils.py`? That's where I'm seeing three other registration funcs. I also see `netbox/plugins/registration.py`, which contains other registrations for plugins. Otherwise, it seems like app registry functions seems to be scattered around a bit.

jeremystretch replied:
Sorry, I meant [`netbox.plugins.registration`](https://github.com/netbox-community/netbox/blob/main/netbox/netbox/plugins/registration.py), where we have similar functions like `register_graphql_schema()` and `register_menu()`. IMO we should go ahead and make this part of the plugins API. It won't be a mandatory change, but plugin authors can opt to register their filtersets now to leverage the new functionality.

jnovinger committed:
Include MODIFIER_EMPTY_FALSE/_TRUE in __all__

Co-authored-by: Jeremy Stretch <jstretch@netboxlabs.com>

jnovinger committed:
Removed explicit checks against QueryField and [Null]BooleanField

I did add them to FORM_FIELD_LOOKUPS, though, to underscore that they
were considered and are intentially empty for future devs.

jnovinger committed:
Switch to sentence case for filter pill text

jnovinger committed:
Fix applied_filters template tag to use field-type-specific lookup labelsresolves

E.g. resolves gt="after" for dates vs "greater than" for numbers

jnovinger commented on netbox/utilities/templatetags/helpers.py:
```diff
@@ -412,28 +414,67 @@ def applied_filters(context, model, form, query_params):
 
+    # Build lookup labels from FORM_FIELD_LOOKUPS
+    lookup_labels = {}
+    for field_lookups in FORM_FIELD_LOOKUPS.values():
```
Yeah, that feels like a housekeeping thing that we can get in v4.5 in a separate issue.

jnovinger replied:
I did make an attempt to make sure each filter gets the right description here: [`11dbe0d` (#20747)](https://github.com/netbox-community/netbox/pull/20747/commits/11dbe0df955d3f6ada54060351c26f5709b87df5)

jnovinger committed:
Verifies that filter pills for exact matches (no lookup
Add test for exact lookup filter pill rendering

jnovinger committed:
Add guard for FilterModifierWidget with no lookups

jnovinger committed:
Remove comparison symbols from numeric filter labels

jnovinger committed:
Match complete tags in widget rendering test assertions

jnovinger committed:
Check all expected lookups in field enhancement tests

jnovinger committed:
Move register_filterset to netbox.plugins.registration

jnovinger committed:
Require registered filterset for filter modifier enhancements

Updates FilterModifierMixin to only enhance form fields when the
associated model has a registered filterset. This provides plugin
safety by ensuring unregistered plugin filtersets fall back to
simple filters without lookup modifiers.

Test changes:
- Create TestModel and TestFilterSet using BaseFilterSet for
automatic lookup generation
- Import dcim.filtersets to ensure Device filterset registration
- Adjust tag field expectations to match actual Device filterset
(has exact/n but not empty lookups)

jnovinger committed:
Attempt to resolve static conflicts

jeremystretch commented:
I'm going to walk back my comment [above](https://github.com/netbox-community/netbox/pull/20747#discussion_r2578249056). I realized that all the registration functions within the `netbox.plugins.registration` module apply specifically to the `plugins` registry. Other registration functions (e.g. `register_model_view()`) are scattered elsewhere. So I think your initial instinct to have `register_filterset()` live in `utilities.filtersets` was correct.

jeremystretch committed:
Move register_filterset() back to utilities.filtersets

jeremystretch committed:
Add register_filterset() to plugins documentation for filtersets

jeremystretch committed:
Reorder import statements
